### ISSUE TITLE
Alternative way to control playback on spotifyd

### ISSUE THREAD

**Issue by joshas** [Spotifyd/spotifyd#432]
I've set up spotifyd on RaspberryPi and it works great as Spotify Connect device. Now I want to add playback controls directly on RaspberryPi, like physical buttons.
Is there an alternative way to control spotifyd: play/pause music, skip to next track, or event start playing another track by giving Spotify URI? I know about DBUS MPRIS, but dbus-send doesn't want to work on Raspbian Lite as it runs without X11 server. Also provided "slim" build of spotifyd ARMv6 probably doesn't include dbus_mpris feature.

**Comment by Yarn** [comment]
Currently I don't think so.

Related: https://github.com/librespot-org/librespot/pull/376

I think it would be useful to implement this in spotifyd (and agree it doesn't belong in librespot).
It would probably be nice to add http endpoints so something simple like a play/pause button doesn't need to worry about having a websocket client.

I can work on this at some point soon probably.

**Comment by SirWindfield** [comment]
Should I assign you to the issue @Yarn?

**Comment by Yarn** [comment]
> Should I assign you to the issue @Yarn?

Sure

**Comment by SirWindfield** [comment]
I am not against this feature. The only real downside I could think of is the fact that this introduces another layer of code that needs to be maintained.

I would like to discuss this feature once you want to start and implement this. Mainly about how to design the web API. It should, nevertheless, definitely be behind a feature flag.

The simplest thing to do would be to map properties and methods in a similar fashion to `http://localhost:3030/propertyname` or `http://localhost:3030/methodname`.

I'm open for other suggestions. The scheme is probably easy to change anyway afterwards as it's just different hard-coded strings.

**Comment by Yarn** [comment]
> The scheme is probably easy to change anyway afterwards

The issue is more breaking compatibility with anything people have already built around it. I think just saying it's unstable at the start is fine though.

**Comment by bluepuma77** [comment]
Sounds like a great idea, using the D-Bus interface isn't my first choice either :-)

I am currently building a music box for my kids, based on a headless Raspberry Pi Zero. I have been researching a while about Spotify players, connect players and controls. My current status is that I need something like `spotifyd` based on [librespot](https://github.com/librespot-org/librespot) and a different tool like [spotipy](https://github.com/plamere/spotipy) or [pyfy](https://github.com/omarryhan/pyfy) (both with different auth mechanisms) to access the Spotify Web API to control it.

I thought about it and here are 3 options I can see:

1.  Re-use the current D-Bus control commands with a different interface technology, e.g. http or Socket.io. For me that enables easier developer access to the existing commands. 
For me the question is if you can provide nice control options like getting current position (album/title/sec) and resuming at position (album/title/sec) - this is what I like to use, besides regular prev/next and search & play title/album.
2. The [Spotify web API](https://developer.spotify.com/documentation/web-api/reference/object-model/) offers a lot of functionality, maybe it makes sense to just provide a http interface and pass back and forth pure REST JSON calls, just adding the necessary authentication. 
This would enable the latest features and you don't have to adapt all the time, it even enables latest beta functionality, like fetching current position in an album and also resuming within an album at the same point later on.
3. If you don't want to be an auth-adding proxy, maybe just provide the current auth credentials to developers. Something like 'spotifyd -viewauth' could just output the current token and the developer can use it to interface with the [Spotify Web API](https://developer.spotify.com/documentation/web-api/reference/object-model/) directly. Kind of a quick&dirty hack, too bad I don't know Rust ;-)

Just my two cents.

**Comment by bluepuma77** [comment]
I found a `librespot` [example](https://github.com/librespot-org/librespot/blob/dev/examples/play.rs) that starts playing a Spotify title. It would be great if that could be exposed somehow. Play title, album, playlist. Trigger start, stop, pause, toggle, next, previous.

**Comment by bluepuma77** [comment]
I found a `librespot` [example](https://github.com/librespot-org/librespot/blob/dev/examples/play.rs) that starts playing a Spotify title. It would be great if that could be exposed somehow. Play title, album, playlist. Trigger start, stop, pause, toggle, next, previous.

Maybe it doesn't even need a http or socket interface but the `spotifyd` CLI can just "talk" to the local `spotifyd` backend. As an example `mocp --server` starts the local backend and you can control it with `mocp --append <folder-with-MP3s> --play`.

**Comment by Yarn** [comment]
I've managed to bootstrap together an initial prototype. It can play/pause when it receives an HTTP request.

I'm currently leaning toward splitting this functionality into a different crate just to keep different versions of things separate (especially with cargo doc where it just overwrites all but one version of a crate). I don't think I'll have much luck actually getting a web server working within the current async code so I'm going with a separate executor communicating with a channel then waking the event loop.

**Comment by intika** [comment]
here is an alternative to control spotifyd without dbus https://github.com/jingweno/spotctl

**Comment by PaulFinch** [comment]
Actually i managed to use MPRIS dbus to control the player. I created a dedicated bus with systemd, and pointed each daemon (spotifyd and my control  script) to the dedicated session dbus. Would work with playerctl as well by exposing the dbus address. I can post more details if needed. The only thing is i needed to compile spotifyd on rpi0 as the binary armv6/full was not available. Took 5 hours ... Anybody know why the armv6 full is not available ?

**Comment by intika** [comment]
> Anybody know why the armv6 full is not available ?
because it takes 5h to build lol :p

**Comment by PaulFinch** [comment]
Indeed, but it should be good reason to avoid people losing 5 hours ^^

**Comment by PaulFinch** [comment]
More details [here](https://www.databullsheet.com/Projects/Raspberry_Pi/Orange_Radio_Upgrade/)

**Comment by intika** [comment]
very nice home made setup, i love it.
 
I just finished an application to control spotify/spotifyd in python... can be helpful for spotifyd's users

https://github.com/Intika-Linux-Spotify/Spotimaster

![Alt text](https://raw.githubusercontent.com/Intika-Linux-Spotify/Spotimaster/master/screens-1.png?raw=true "Capture")

**Comment by PaulFinch** [comment]
What a good proper work.

**Comment by Yarn** [comment]
bad bot, i'm just extremely slow

**Comment by schickling** [comment]
Still interested in this. Please re-open.
